### PR TITLE
[DO NOT MERGE, BUT DO REVIEW!] HF x PyTorch DDP Blog

### PR DESCRIPTION
**Pull request by muellerzr**
This is a **draft** collaboration blog between Hugging Face and PyTorch on Distributed Data Parallelism. It doesn't 100% fit all of the usual blog styling nits that HF blogs would because we are still looking into whether it will be published only on PyTorch, HF then linked, etc. So for now it doesn't have any of our custom watermarkings etc. 

Sending this off to review with our team before passing it along further to the folks at torch to have their crack at edits etc. Want to make sure it's well and polished before then 😉

### CODE REVIEW

File: pytorch-ddp-accelerate-transformers.md [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+            output = model(data)
+            pred = output.argmax(dim=1, keepdim=True)
+            correct += pred.eq(target.view_as(pred)).sum().item()
+    print(f'Accuracy: {100. * correct / len(test_loader.dataset)}')
+```
+
+Next let's talk about how Accelerate can help. There's a few issues with the above code:
+
+1. This is slightly inefficient, given that `n` dataloaders are made based on each device and pushed.
+2. This code will **only** work for multi-GPU, so special care would need to be made for it to be ran on a single node again, or on TPU.
```
Comment (sgugger) [inline pytorch-ddp-accelerate-transformers.md:231]: Ah, it's mentioned here. Might be enough!

File: pytorch-ddp-accelerate-transformers.md [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+
+Accelerate helps this through the [`Accelerator`](https://huggingface.co/docs/accelerate/v0.12.0/en/package_reference/accelerator#accelerator) class. Through it, the code remains much the same except for three lines of code when comparing a single node to multinode, as shown below:
+
+```python
+def train_ddp_accelerate():
+    accelerator = Accelerator(device_placement=True)
```
Comment (sgugger) [inline pytorch-ddp-accelerate-transformers.md:237]:
```suggestion
    accelerator = Accelerator()
```
This is the default.

File: pytorch-ddp-accelerate-transformers.md [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+
+This wraps as much training as possible while still being able to train on distributed systems without the user needing to do anything at all.
+
+First we need to import the Trainer:
+
+```python
+from transformers import Trainer
+```
+
+Then we define some TrainingArguments since it will handle the dataloader creation. The trainer also works through dictionaries, so a custom colalte function needs to be made.
```
Comment (sgugger) [inline pytorch-ddp-accelerate-transformers.md:322]:
```suggestion
Then we define some `TrainingArguments` to control all the usual hyper-parameters. The trainer also works through dictionaries, so a custom collate function needs to be made.
```

**Review by sgugger — APPROVED** [review]
Thanks for drafting this! Left a couple of comments.

File: pytorch-ddp-accelerate-transformers.md (hunk 2 above)
Comment (muellerzr) [reply pytorch-ddp-accelerate-transformers.md:237]: It is, but also in the docs where we assume no prior knowledge (in our docs) we have `device_placement=True`. 

Leave the default and make a note about device placement probably then?

Comment (sgugger) [reply pytorch-ddp-accelerate-transformers.md:237]: Yeah, we can also update the docs since it was originally set like this to show how to remove the automatic device placement but no one ever complained about it.

File: pytorch-ddp-accelerate-transformers.md [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+    dist.destroy_process_group()
+```
+
+The last piece of the puzzle is *how do I send my data and model to another GPU?*
+
+This is where the `DistributedDataParallel` module comes into play. It will copy your model into each GPU, and when `loss.backward()` is called the gradients across all these copies of the model will be averaged, backprop is performed, and each of the weights get updated to have the same weights across all devices.
+
+Below is an example of our training setup, refactored as a function, with this capability:
+
+> Note: Here rank is the overall rank of the GPU based on all other devices, meaning they have a rank of `0 -> n`
```
[the remark this answers is not in the document]
Comment (muellerzr) [reply pytorch-ddp-accelerate-transformers.md:144]: Rewording more accurately:

> Note: Here rank is the overall rank of the current GPU compared to all the other GPUs available, meaning they have a rank of `0 -> n-1`

File: pytorch-ddp-accelerate-transformers.md [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+
+The optimizer needs to be declared based on the model *on the specific device* for all of the gradients to properly be calculated.
+
+Lastly, to run the script PyTorch has a convient `torchrun` command line module that can help. Just pass in the number of nodes it should use as well as the script to run and you are set:
+
+```bash
+torchrun --nproc_per_nodes=2 --nnodes=1 example_script.py
+```
+
+The above will run the training script on two GPUs that live on a single machine and this is the barebones for performing distributed training with PyTorch.
```
[the remark this answers is not in the document]
Comment (muellerzr) [reply pytorch-ddp-accelerate-transformers.md:174]: Went in the middle with the below clarification, phrasing as:
"The above will run the training script on two GPUs that live on a single machine and this is the barebones for performing only distributed training with PyTorch."

(Aka adding **only**)

File: pytorch-ddp-accelerate-transformers.md [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+    dist.init_process_group("gloo", rank=rank, world_size=world_size)
+
+def cleanup():
+    "Cleans up the distributed environment"
+    dist.destroy_process_group()
+```
+
+The last piece of the puzzle is *how do I send my data and model to another GPU?*
+
+This is where the `DistributedDataParallel` module comes into play. It will copy your model into each GPU, and when `loss.backward()` is called the gradients across all these copies of the model will be averaged, backprop is performed, and each of the weights get updated to have the same weights across all devices.
```
Comment (pacman100) [inline pytorch-ddp-accelerate-transformers.md:140]:
```suggestion
This is where the `DistributedDataParallel` module comes into play. It will copy your model onto each GPU, and when `loss.backward()` is called the backpropagation is performed and the resulting gradients across all these copies of the model will be averaged/reduced. This ensures each device has the same weights post the optimizer step.
```

File: pytorch-ddp-accelerate-transformers.md [hunk 7]
```diff
@@ -0,0 +1,427 @@
...
+    model.train()
+    for batch_idx, (data, target) in enumerate(train_loader):
+        data, target = data.to(device), target.to(device)
+        output = model(data)
+        loss = F.nll_loss(output, target)
+        loss.backward()
+        optimizer.step()
+        optimizer.zero_grad()
+    cleanup()
+```
```
Comment (pacman100) [inline pytorch-ddp-accelerate-transformers.md:164]: I believe changes that are necessary for dataloaders are left out. That is a significant detail and quite a bit of a hassle.  Accelerate abstracts and that is great for users. So, good to detail this too. For normal datasets, that would mean using the `DistributedSampler` sampler https://pytorch.org/docs/stable/data.html#torch.utils.data.distributed.DistributedSampler.

File: pytorch-ddp-accelerate-transformers.md [hunk 8]
```diff
@@ -0,0 +1,427 @@
...
+    train_loader, test_loader, model, optimizer = accelerator.prepare(
+        train_loader, test_loader, model, optimizer
+    )
+
+    # Train for a single epoch
+    model.train()
+    for batch_idx, (data, target) in enumerate(train_loader):
+        output = model(data)
+        loss = F.nll_loss(output, target)
+        loss.backward()
```
Comment (pacman100) [inline pytorch-ddp-accelerate-transformers.md:266]:
```suggestion
        accelerator.backward()
```

File: pytorch-ddp-accelerate-transformers.md [hunk 9]
```diff
@@ -0,0 +1,427 @@
...
+        optimizer.step()
+        optimizer.zero_grad()
+    
+    # Evaluate
+    model.eval()
+    correct = 0
+    with torch.no_grad():
+        for data, target in test_loader:
+            data, target = data.to(device), target.to(device)
```
Comment (pacman100) [inline pytorch-ddp-accelerate-transformers.md:275]:
```suggestion
```

File: pytorch-ddp-accelerate-transformers.md [hunk 10]
```diff
@@ -0,0 +1,427 @@
...
+    for batch_idx, (data, target) in enumerate(train_loader):
+        data, target = data.to(device), target.to(device)
+        output = model(data)
+        loss = F.nll_loss(output, target)
+        loss.backward()
+        optimizer.step()
+        optimizer.zero_grad()
+    cleanup()
+```
+
```
Comment (pacman100) [inline pytorch-ddp-accelerate-transformers.md:165]: Also, details on handling outputs during evaluation, i.e, gathering from all the processes is also quite a bit of hassle and an important detail when woking in distributed setting else evaluations won't be reliable hampering the experimentation process. Accelerate abstracts this too making life simple for users.
Without Accelerate one would have code like below:
```
gather_predictions = [predictions.clone() for _ in range(dist.get_world_size())]
torch.distributed.all_gather(gather_predictions, predictions)
batch_predictions =  torch.cat(gather_predictions, dim=0)

gather_labels = [batch["labels"].clone() for _ in range(dist.get_world_size())]
torch.distributed.all_gather(gather_labels, batch["labels"])
batch_labels =  torch.cat(gather_labels, dim=0)

metric.add_batch(
    predictions=batch_predictions.detach().cpu().numpy(),
    references=batch_labels.detach().cpu().numpy(),
)

```

File: pytorch-ddp-accelerate-transformers.md [hunk 11]
```diff
@@ -0,0 +1,427 @@
...
+torchrun --nproc_per_nodes=2 --nnodes=1 example_script.py
+```
+
+The above will run the training script on two GPUs that live on a single machine and this is the barebones for performing only distributed training with PyTorch.
+
+Now let's talk about Accelerate, a library aimed to make this process more seameless and also help with a few best practices
+
+## 🤗 Accelerate
+
+[Accelerate](https://huggingface.co/docs/accelerate) is a library designed to allow you to perform what we just did above, without needing to modify your code greatly. On top of this, the data pipeline innate to Accelerate can also improve performance to your code as well.
```
Comment (pacman100) [inline pytorch-ddp-accelerate-transformers.md:180]: I believe mentioning the abstractions for dataloaders and metrics gathering is also important in addition to support for various device types and model related changes in distributed setup.

File: pytorch-ddp-accelerate-transformers.md [hunk 12]
```diff
@@ -0,0 +1,427 @@
...
+            data, target = data.to(device), target.to(device)
+            output = model(data)
+            pred = output.argmax(dim=1, keepdim=True)
+            correct += pred.eq(target.view_as(pred)).sum().item()
+    print(f'Accuracy: {100. * correct / len(test_loader.dataset)}')
+```
+
+Now any and all custom code needed to launch your code on any distributed setup is simplified to the `Accelerator` object. This code can then still be ran through the `torchrun` CLI, or through Accelerate's own CLI interface, [`accelerate launch`](https://huggingface.co/docs/accelerate/v0.12.0/en/basic_tutorials/launch).
+
+As a result its now trivialized to perform distributed training with Accelerate and keeping as much of the barebones PyTorch code the same as possible.
```
Comment (pacman100) [inline pytorch-ddp-accelerate-transformers.md:284]: Needs rewording

File: pytorch-ddp-accelerate-transformers.md [hunk 13]
```diff
@@ -0,0 +1,427 @@
...
+
+Earlier it was mentioned that Accelerate also makes the DataLoaders more efficient. This is through custom Samplers that will send parts of the batches automatically to different devices during training allowing for a single copy of the data to be known at one time, rather than four at once into memory.
```
Comment (pacman100) [inline pytorch-ddp-accelerate-transformers.md:286]: I think here the important detail is that the user doesn't have to worry about making sure each device has a subset of the original dataset that is exclusive to it. Everything is handled under the hood. Also, single copy vs multiple copies isn't accurate because `DataLoaderDispatcher` (in case of IterableDataset) sends num_processes*batch_size data from rank 0 to all ranks leading to data deduplication/multiple copies wherein each rank then picks the slice exclusive to it.

**Review by pacman100 — COMMENTED** [review]
Thank you @muellerzr for the insightful blog 🤗. Please find my comments and suggestions.

File: pytorch-ddp-accelerate-transformers.md [hunk 14]
```diff
@@ -0,0 +1,427 @@
...
+        data, target = data.to(device), target.to(device)
+        output = model(data)
+        pred = output.argmax(dim=1, keepdim=True)
+        correct += pred.eq(target.view_as(pred)).sum().item()
+print(f'Accuracy: {100. * correct / len(test_loader.dataset)}')
+```
+
+Typically from here, one could either throw all of this into a python script or run it on a Jupyter Notebook.
+
+However, how would you then get this to run on say two GPUs if these resources available? Just doing `python myscript.py` will only ever run the script using a single GPU. This is where `torch.distributed` comes into play
```
[the remark this answers is not in the document]
Comment (muellerzr) [reply pytorch-ddp-accelerate-transformers.md:109]: Met in the middle on this suggestion and reworded it some

**Comment by muellerzr** [comment]
Going ahead with the blogpost to have it launch by Friday. Still waiting on the torch folks but I think we're set to go from what I can see :) 

We'll split up Sourab's comments into a more advanced "Part 2" to leave this one as intro friendly.

@LysandreJik can you give it a glance and then I think we'll be set :) Thanks!

File: pytorch-ddp-accelerate-transformers.md (hunk 11 above)
Comment (muellerzr) [reply pytorch-ddp-accelerate-transformers.md:197]: Will do this in a followup as a deep dive, so that this can be surface level :)

File: pytorch-ddp-accelerate-transformers.md (hunk 12 above)
Comment (muellerzr) [reply pytorch-ddp-accelerate-transformers.md:301]: Fixed :)

File: pytorch-ddp-accelerate-transformers.md [hunk 15]
```diff
@@ -0,0 +1,444 @@
...
+
+> Note: Here rank is the overall rank of the current GPU compared to all the other GPUs available, meaning they have a rank of `0 -> n-1`
+
+```python
+from torch.nn.parallel import DistributedDataParallel as DDP
+
+def train(model, rank, world_size):
+    setup(rank, world_size)
+    model = model.to(rank)
+    ddp_model = DDP(model, device_ids=[rank])
```
Comment (mrshenli) [inline pytorch-ddp-accelerate-transformers.md:169]: Thanks @muellerzr for putting up the blog post!!

I just wanna mention a few more techniques to further speed up DDP. No need to change version if the goal is not squeezing out the last bit of perf from the cluster. 

1. Setting `gradient_as_bucket_view` to `True`. This can make `param.grad` points to an offset of DDP's bucket, which can help save two copies (i.e., grad to bucket, bucket to grad).
2.  It is possible to overlap optimizer step with DDP's backward communication by using the DDP comm hook. (https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel.register_comm_hook) However, this will only make a difference if communication delay is dominating.

Comment (muellerzr) [reply pytorch-ddp-accelerate-transformers.md:169]: Thanks @mrshenli! The goal of this one is not, but I'll definitely include it in the next! (And look at having this happen also in `accelerate`), thanks for the TIL! 🤗
